datafusion/core/src/physical_optimizer/sort_enforcement2.rs

```diff
@@ -0,0 +1,2872 @@
+    /// Parent has the SinglePartition requirement to children
+    satisfy_single_distribution: bool,
+    /// Parent required sort ordering
+    required_ordering: Option<Vec<PhysicalSortRequirements>>,
+    /// The adjusted request sort ordering to children.
+    /// By default they are the same as the plan's required input ordering, but can be adjusted based on parent required sort ordering properties.
+    adjusted_request_ordering: Vec<Option<Vec<PhysicalSortRequirements>>>,
+}
+
+impl PlanWithSortRequirements {
+    pub fn init(plan: Arc<dyn ExecutionPlan>) -> Self {
+        let impact_result_ordering = plan.output_ordering().is_some()
+            || plan.output_partitioning().partition_count() <= 1
+            || plan.as_any().downcast_ref::<GlobalLimitExec>().is_some()
+            || plan.as_any().downcast_ref::<LocalLimitExec>().is_some();
+        let request_ordering = plan.required_input_ordering();
+        PlanWithSortRequirements {
+            plan,
+            impact_result_ordering,
+            satisfy_single_distribution: false,
+            required_ordering: None,
+            adjusted_request_ordering: request_ordering,
+        }
+    }
+
+    pub fn new_without_impact_result_ordering(plan: Arc<dyn ExecutionPlan>) -> Self {
+        let request_ordering = plan.required_input_ordering();
+        PlanWithSortRequirements {
+            plan,
+            impact_result_ordering: false,
+            satisfy_single_distribution: false,
+            required_ordering: None,
+            adjusted_request_ordering: request_ordering,
+        }
+    }
+
+    pub fn children(&self) -> Vec<PlanWithSortRequirements> {
+        let plan_children = self.plan.children();
+        assert_eq!(plan_children.len(), self.adjusted_request_ordering.len());
+
+        izip!(
+            plan_children.into_iter(),
+            self.adjusted_request_ordering.clone().into_iter(),
+            self.plan.maintains_input_order().into_iter(),
+            self.plan.required_input_distribution().into_iter(),
+        )
+        .map(
+            |(child, from_parent, maintains_input_order, required_dist)| {
+                let child_satisfy_single_distribution =
+                    matches!(required_dist, Distribution::SinglePartition)
+                        || (self.satisfy_single_distribution
+                            && self
+                                .plan
+                                .as_any()
+                                .downcast_ref::<CoalescePartitionsExec>()
+                                .is_none());
+                let child_impact_result_ordering = if self
+                    .plan
+                    .as_any()
+                    .downcast_ref::<GlobalLimitExec>()
+                    .is_some()
+                    || self
+                        .plan
+                        .as_any()
+                        .downcast_ref::<LocalLimitExec>()
+                        .is_some()
```

Comment:
Comment above applies here. Also you can construct a util function `is_limit` to check for Exec is `GlobalLimitExec ` or `LocalLimitExec`. This util can be used for the check at the `PlanWithSortRequirements::init`.

Reply:
👍